controllers/cas.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	//find the token
+	if ok, response, issuedService := object.GetCasAuthenticationSuccessByTicket(ticket); ok {
+
+		//check whether service is the one for which we previously issued token
+		if strings.HasPrefix(service, issuedService) {
+			serviceResponse.Success = response
+		} else {
+			//service not match
+			c.sendCasAuthenticationResponseErr(INVALID_SERVICE, fmt.Sprintf("service %s and %s does not match", service, issuedService), format)
+
+			return
+		}
+	} else {
+		//token not found
+		c.sendCasAuthenticationResponseErr(INVALID_TICKET, fmt.Sprintf("Ticket %s not recognized", ticket), format)
+		return
+	}
+
+	if pgtUrl != "" && serviceResponse.Failure == nil {
+		//that means we are in proxy web flow
+		pgt := object.StoreCasAuthenticationSuccessForPgt(serviceResponse.Success, service)
+		pgtiou := serviceResponse.Success.ProxyGrantingTicket
+		//todo: check whether it is https
+		pgtUrlObj, err := url.Parse(pgtUrl)
+		if pgtUrlObj.Scheme != "https" {
+			c.sendCasAuthenticationResponseErr(INVALID_PROXY_CALLBACK, "callback is not https", format)
+			return
+		}
+		//make a request to pgturl passing pgt and pgtiou
+		if err != nil {
+			c.sendCasAuthenticationResponseErr(INTERNAL_ERROR, err.Error(), format)
+			return
+		}
+		param := pgtUrlObj.Query()
+		param.Add("pgtId", pgt)
+		param.Add("pgtIou", pgtiou)
+		pgtUrlObj.RawQuery = param.Encode()
+
+		request, err := http.NewRequest("GET", pgtUrlObj.String(), nil)
+		if err != nil {
+			c.sendCasAuthenticationResponseErr(INTERNAL_ERROR, err.Error(), format)
+			return
+		}
+
+		resp, err := http.DefaultClient.Do(request)
+		if err != nil || !(resp.StatusCode >= 200 && resp.StatusCode < 400) {
+			//failed to send request
+			c.sendCasAuthenticationResponseErr(INVALID_PROXY_CALLBACK, err.Error(), format)
+			return
+		}
+	}
+	// everything is ok, send the response
+	if format == "json" {
+		c.Data["json"] = serviceResponse
+		c.ServeJSON()
+	} else {
+		c.Data["xml"] = serviceResponse
+		c.ServeXML()
+	}
+	fmt.Println(serviceResponse)
+}
+
+func (c *RootController) Proxy() {
+	pgt := c.Input().Get("pgt")
+	targetService := c.Input().Get("targetService")
+	format := c.Input().Get("format")
+	if pgt == "" || targetService == "" {
+		c.sendCasProxyResponseErr(INVALID_REQUEST, "pgt and targetService must be all presented", format)
+		return
+	}
+
+	ok, authenticationSuccess, issuedService := object.GetAuthenticationSuccessByPgt(pgt)
+	if !ok {
+		c.sendCasProxyResponseErr(UNAUTHORIZED_SERVICE, "service not authorized", format)
+		return
+	}
+
+	newAuthenticationSuccess := authenticationSuccess.DeepCopy()
+	if newAuthenticationSuccess.Proxies == nil {
+		newAuthenticationSuccess.Proxies = &object.CasProxies{}
+	}
+	newAuthenticationSuccess.Proxies.Proxies = append(newAuthenticationSuccess.Proxies.Proxies, issuedService)
+	proxyTicket := object.StoreCasAuthenticationSuccessForProxyTicket(&newAuthenticationSuccess, targetService)
+
+	var serviceResponse object.CasServiceResponse
```

Comment:
Don't use "var xxx xxx" to define object again. Use:

![image](https://user-images.githubusercontent.com/3787410/161295929-7bb0afdb-7249-4759-89a9-8ddaa896d3da.png)